src/Ui/Table.elm

```diff
@@ -0,0 +1,464 @@
+module Ui.Table exposing
+    ( ColumnAlignment(..)
+    , ColumnConfig
+    , Flex(..)
+    , Model
+    , Msg
+    , SortDirection(..)
+    , TableConfig
+    , addColumn
+    , addExtendableView
+    , cellColumns
+    , cellCustom
+    , cellFloat
+    , cellInt
+    , cellLink
+    , cellRows
+    , cellText
+    , column
+    , columnWithDataEntryId
+    , defaultConfig
+    , hideColumn
+    , setColumnAlignment
+    , setColumnWidth
+    , showColumn
+    , update
+    , view
+    )
+
+import Css
+import Css.Transitions exposing (transition)
+import Dict exposing (Dict)
+import Html.Styled exposing (Attribute, Html, a, div, span)
+import Html.Styled.Attributes exposing (css, href)
+import Html.Styled.Events exposing (onClick)
+import Html.Styled.Keyed exposing (lazyNode3)
+import List
+import Maybe.Extra exposing (isJust)
+import Ui.Css.Palette as Palette
+import Ui.Css.TextStyle exposing (toCssStyle)
+import Ui.Icon
+import Ui.Palette
+import Ui.Text as Text
+import Ui.TextStyle as TextStyle exposing (TextStyle)
+
+
+type SortDirection
+    = Ascending
+    | Descending
+
+
+type ColumnAlignment
+    = Left
+    | Center
+    | Right
+
+
+type ColumnContent msg
+    = ColumnContent (Html msg)
+
+
+type alias ColumnConfig record msg =
+    { name : String
+    , width : Css.LengthOrAuto Css.Px
+    , content : DataEntryId -> record -> ColumnContent msg
+    , visible : Bool
+    , alignment : ColumnAlignment
+    , weight : Int
+    }
+
+
+type alias TableConfig record msg =
+    { columns : Dict ColumnId (ColumnConfig record msg)
+    , onHeaderClick : Maybe (ColumnId -> msg)
+    , extendableView : Maybe (record -> Html msg)
+    , liftMsg : Msg -> msg
+    }
+
+
+type alias Model record =
+    { sortDirection : SortDirection
+    , sortIndex : String
+    , data : Dict DataEntryId ( record, Bool )
+    }
+
+
+type alias ColumnId =
+    String
+
+
+type alias DataEntryId =
+    String
+
+
+type Msg
+    = ToggleAccordionView DataEntryId
+
+
+update : Msg -> Model v -> ( Model v, Cmd Msg )
+update msg model =
+    case msg of
+        ToggleAccordionView dId ->
+            ( { model
+                | data =
+                    Dict.update dId
+                        (Maybe.map <|
+                            \entry ->
+                                ( Tuple.first entry
+                                , not <| Tuple.second entry
+                                )
+                        )
+                        model.data
+              }
+            , Cmd.none
+            )
+
+
+defaultConfig : (Msg -> msg) -> TableConfig record msg
+defaultConfig liftMsg =
+    { onHeaderClick = Nothing
+    , columns = Dict.empty
+    , extendableView = Nothing
+    , liftMsg = liftMsg
+    }
+
+
+addColumn :
+    ( String, Int -> ColumnConfig record msg )
```

Comment:
should be ColumnId. Use "newtypes" to make it less easy to forget to use domain names?

Reply:
or not, but make sure to update it

Reply:
sure